Serialization bug with pack indexing

The following test case fails with clang trunk:

```cpp

// RUN: rm -rf %t
// RUN: split-file %s %t

// RUN: %clang_cc1 -std=c++2c -x c++-header %t/test.h -emit-pch -o %t/test.pch
// RUN: %clang_cc1 -std=c++2c -x c++ %t/test.cpp -include-pch %t/test.pch

//--- test.h
template <class... Ts> struct _TypeInfo {
	template <int id> using type = Ts...[id];
	static constexpr int count = sizeof...(Ts);
};
using TypeInfo = _TypeInfo<int>;

TypeInfo::type<0> a;

//--- test.cpp
int main() {
  TypeInfo::type<0> a;
}


```

I can see false positive diagnostics:

```
llvm-project-Build/Build/tools/clang/test/PCH/Output/pack-indexing-2.cpp.tmp/test.h:2:39: error: invalid index 0 for pack 'Ts' of size 0
# |     2 |         template <int id> using type = Ts...[id];
# |       |                                              ^
```

This was first reported in https://github.com/clangd/clangd/issues/2529

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -606,6 +606,7 @@ Bug Fixes to C++ Support
 - Fix a crash when extracting unavailable member type from alias in template deduction. (#GH165560)
 - Fix incorrect diagnostics for lambdas with init-captures inside braced initializers. (#GH163498)
 - Fixed an issue where templates prevented nested anonymous records from checking the deletion of special members. (#GH167217)
+- Fixed a serialization bug in PackIndexingType, where we failed to expand those packs from a PCH/module. (#GH172464)
 - Fixed spurious diagnoses of certain nested lambda expressions. (#GH149121) (#GH156579)
 - Fix the result of ``__is_pointer_interconvertible_base_of`` when arguments are qualified and passed via template parameters. (#GH135273)
 

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -606,7 +606,7 @@ Bug Fixes to C++ Support
 - Fix a crash when extracting unavailable member type from alias in template deduction. (#GH165560)
 - Fix incorrect diagnostics for lambdas with init-captures inside braced initializers. (#GH163498)
 - Fixed an issue where templates prevented nested anonymous records from checking the deletion of special members. (#GH167217)
-- Fixed a serialization bug in PackIndexingType, where we failed to expand those packs from a PCH/module. (#GH172464)
+- Fixed serialization of pack indexing types, where we failed to expand those packs from a PCH/module. (#GH172464)
 - Fixed spurious diagnoses of certain nested lambda expressions. (#GH149121) (#GH156579)
 - Fix the result of ``__is_pointer_interconvertible_base_of`` when arguments are qualified and passed via template parameters. (#GH135273)
 

--- a/clang/include/clang/AST/TypeProperties.td
+++ b/clang/include/clang/AST/TypeProperties.td
@@ -486,9 +486,12 @@ let Class = PackIndexingType in {
   def : Property<"isFullySubstituted", Bool> {
     let Read = [{ node->isFullySubstituted() }];
   }
+  def : Property<"expansions", Array<QualType>> {
+    let Read = [{ node->getExpansions() }];
+  }
 
   def : Creator<[{
-    return ctx.getPackIndexingType(pattern, indexExpression, isFullySubstituted);
+    return ctx.getPackIndexingType(pattern, indexExpression, isFullySubstituted, expansions);
   }]>;
 }
 

--- a/clang/test/PCH/pack-indexing-2.cpp
+++ b/clang/test/PCH/pack-indexing-2.cpp
@@ -1,0 +1,18 @@
+// RUN: rm -rf %t
+// RUN: split-file %s %t
+
+// RUN: %clang_cc1 -std=c++2c -x c++-header %t/GH172464.h -emit-pch -o %t/GH172464.pch
+// RUN: %clang_cc1 -std=c++2c -x c++ %t/GH172464.cpp -include-pch %t/GH172464.pch
+
+//--- GH172464.h
+template <class... Ts> struct _TypeInfo {
+	template <int id> using type = Ts...[id];
+};
+using TypeInfo = _TypeInfo<int>;
+
+TypeInfo::type<0> a;
+
+//--- GH172464.cpp
+int main() {
+  TypeInfo::type<0> a;
+}

